File: solutions/0020. Valid Parentheses/0020.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
public:
	bool isValid(const string &s) {
		stack <char, vector<char>> st;
		
		for (char c : s) {
			if (c == '(') {
				st.push(')');
			} else if (c == '{') {
				st.push('}');
			} else if (c == '[') {
				st.push(']');
			} else {
				if (st.empty() || st.top() != c) {
					return false;
				}
				st.pop();
			}
		}
		return st.empty();
	}
};

File: solutions/0020. Valid Parentheses/0020.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def isValid(self, s: str) -> bool:
        st = []

        for c in s:
            if c == '(':
                st.append(')')
            elif c == '{':
                st.append('}')
            elif c == '[':
                st.append(']')
            elif not st or st.pop() != c:
                return False

        return not st